[lldb] Port Process::PrintWarning* to use the new diagnostic events

JDevlieghere opened the pull request:
Port the two Process::PrintWarning functions to use the new diagnostic
events through Debugger::ReportWarning. I kept the wrapper function in
the process, but delegated the work to the Module. Consistent with the
current code, the Module ensures the warning is only printed once per
module.

Differential revision: https://reviews.llvm.org/D123698

JDevlieghere committed:
[lldb] Port Swift Process::PrintWarning* to use the new diagnostic events

JDevlieghere commented:
@swift-ci please test

kastiglione commented on lldb/source/Core/Module.cpp:
```diff
@@ -1156,6 +1156,64 @@ bool Module::FileHasChanged() const {
+        swift::version::Version::getCurrentCompilerVersion();
+    if (sym_file_version != swift_version) {
+      StreamString ss;
+      ss << GetFileSpec().GetFilename()
+         << " was compiled with a different Swift compiler "
+            "(version '"
+         << sym_file_version.getAsString()
+         << "') than the Swift compiler integrated into LLDB "
+            "(version '"
+         << swift_version.getAsString()
+         << "'). Swift expression evaluation requires a matching "
+            "compiler and debugger from the same toolchain.";
```
This would probably read better, and grep better, expressed as a "format" string or printf.

kastiglione commented on lldb/include/lldb/Core/Module.h:
```diff
@@ -818,6 +818,22 @@ class Module : public std::enable_shared_from_this<Module>,
+      llvm::Optional<lldb::user_id_t> debugger_id = llvm::None);
+  void ReportWarningToolchainMismatch(
+      CompileUnit &comp_unit,
+      llvm::Optional<lldb::user_id_t> debugger_id = llvm::None);
```
Repeating a question for [D123698](https://reviews.llvm.org/D123698), the use of `Optional` and a default is not used in this PR. Will this ever be called without a debugger_id?

kastiglione replied:
This was answered [here](https://reviews.llvm.org/D123698#inline-1185285).

JDevlieghere committed:
[lldb] Port Process::PrintWarning* to use the new diagnostic events

Port the two Process::PrintWarning functions to use the new diagnostic
events through Debugger::ReportWarning. I kept the wrapper function in
the process, but delegated the work to the Module. Consistent with the
current code, the Module ensures the warning is only printed once per
module.

Differential revision: https://reviews.llvm.org/D123698

(cherry picked from commit 99d9c44434f57df5f188146925eea3bd8771260d)

JDevlieghere commented:
@swift-ci please test

JDevlieghere commented:
@swift-ci please test

JDevlieghere commented:
@swift-ci please test macos
